core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -653,7 +653,11 @@ object Stream {
    * Constructs an infinite stream from a `Queue`.
    */
   final def fromQueue[A](queue: Queue[A]): Stream[Nothing, A] =
-    unfoldM(())(_ => queue.take.map(a => Some((a, ()))))
+    unfoldM(())(
+      _ =>
+        queue.take
+          .redeem0(cause => if (cause.interrupted) IO.now(None) else IO.fail0(cause), a => IO.now(Some((a, ()))))
```

Comment:
`redeem0` actually cannot recover from interruption.

It's possible we need to modify the `orElse` combinator so that if the first one is interrupted, it chooses the second. We probably need to guarantee that `race` behaves similarly (I think it does).

Could use a failing test here just so we know when we fix the problem.

Reply:
Right! I even worked on that 🤦‍♂️The comment on `redeem0` is misleading "allows recovering from any kind of failure", I will change it.
I wrote a test that indeed fails. I'll comment it for now.

So the plan is to fix `redeem0` eventually?

Reply:
Or do you mean `orElse` should no longer use `redeem0` and instead something like `fork.observe` to be able to catch anything?

Reply:
At the moment, it can't. In RTS:
```scala
unwindStack(!io.cause.interrupted)
```
and in `unwindStack` (where `catchError = !io.cause.interrupted`):
```scala
case a: IO.Redeem[_, _, _, _] if catchError =>
            errorHandler = a.err.asInstanceOf[Any => IO[Any, Any]]
```
so this test is actually irrelevant.

Reply:
Yeah. What do you think about what the type reflects? redeem0’s type will lead you to believe that you’d be able to observe the interruption.

Reply:
Yes, true. Semantically, interruption is still a bona fide cause for failure. The previous `redeem0` had separate continuations for checked vs unchecked errors, but thanks to the redesign of errors we can now handle both similarly and get more information out of it, so perhaps the advantages are worth this minor impediment?

Reply:
Right. It is minor but ExitResult and Cause perfectly describe how you can observe fiber endings except for this part.

I know you've done a lot of work on that so I hope I'm not barging into an open door.

Perhaps ExitResult should be formulated as `Succeeded a | Interrupted | Failed (Checked e | Unchecked Throwable)`? I can open another issue to discuss this.

Reply:
`Cause` perfectly describes how fibers fail right now, because interruption is a perfectly valid way for a fiber to fail. 

In fact, the situation is more nuanced than this, because `Cause` can be composite from many fibers; some of which may have been interrupted, others of which may have failed for specific reasons. 

Even more complicated, during the process of interruption, finalizers may throw unchecked errors, which means the cause is multi-factoral and must reflect both the interruption and the unchecked errors.

All of this is losslessly and precisely encoded with `Cause`. So it's a good bet that's not changing. 😆 

The only "strange" thing is that when you call `redeem`, you think you _might_ be able to handle `Cause.Interrupted`. But you can't really recover from interruption, by design (it's a simple change to make, but we decided against it). Anywhere else where you see a `Cause`, it's quite possible it's `Cause.Interrupted` (or contains a `Cause.Interrupted`, anyway).